Sources/SPMUtility/dlopen.swift

```diff
@@ -0,0 +1,116 @@
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import SPMLibc
+
+public final class DLHandle {
+  #if os(Windows)
+    typealias Handle = HMODULE
+  #else
+    typealias Handle = UnsafeMutableRawPointer
+  #endif
+    var rawValue: Handle? = nil
+
+    init(rawValue: Handle) {
+        self.rawValue = rawValue
+    }
+
+    deinit {
+        precondition(rawValue == nil, "DLHandle must be closed or explicitly leaked before destroying")
+    }
+
+    public func close() throws {
+        if let handle = rawValue {
+          #if os(Windows)
+            guard FreeLibrary(handle) != 0 else {
+                throw DLError.close("Failed to FreeLibrary: \(GetLastError())")
+            }
+          #else
+            guard dlclose(handle) == 0 else {
+                throw DLError.close(dlerror() ?? "unknown error")
+            }
+          #endif
+        }
+        rawValue = nil
+    }
+
+    public func leak() {
+        rawValue = nil
+    }
+}
+
+public struct DLOpenFlags: RawRepresentable, OptionSet {
+
+  #if !os(Windows)
+    public static let lazy: DLOpenFlags = DLOpenFlags(rawValue: RTLD_LAZY)
+    public static let now: DLOpenFlags = DLOpenFlags(rawValue: RTLD_NOW)
+    public static let local: DLOpenFlags = DLOpenFlags(rawValue: RTLD_LOCAL)
+    public static let global: DLOpenFlags = DLOpenFlags(rawValue: RTLD_GLOBAL)
+
+    // Platform-specific flags.
+  #if os(macOS)
+    public static let first: DLOpenFlags = DLOpenFlags(rawValue: RTLD_FIRST)
+    public static let deepBind: DLOpenFlags = DLOpenFlags(rawValue: 0)
+  #else
+    public static let first: DLOpenFlags = DLOpenFlags(rawValue: 0)
+    public static let deepBind: DLOpenFlags = DLOpenFlags(rawValue: RTLD_DEEPBIND)
+  #endif
+  #endif
+
+    public var rawValue: Int32
+
+    public init(rawValue: Int32) {
+        self.rawValue = rawValue
+    }
+}
+
+public enum DLError: Swift.Error {
+    case `open`(String)
+    case close(String)
+}
+
+public func dlopen(_ path: String?, mode: DLOpenFlags) throws -> DLHandle {
+  #if os(Windows)
+    guard let handle = path?.withCString(encodedAs: UTF16.self, LoadLibraryW) else {
+        throw DLError.open("LoadLibraryW failed: \(GetLastError())")
+    }
+  #else
+    guard let handle = SPMLibc.dlopen(path, mode.rawValue) else {
+        throw DLError.open(dlerror() ?? "unknown error")
+    }
+  #endif
+    return DLHandle(rawValue: handle)
+}
+
+public func dlsym<T>(_ handle: DLHandle, symbol: String) -> T? {
+  #if os(Windows)
+    guard let ptr = GetProcAddress(handle.rawValue!, symbol) else {
+        return nil
+    }
+  #else
+    guard let ptr = dlsym(handle.rawValue!, symbol) else {
+        return nil
+    }
+  #endif
+    return unsafeBitCast(ptr, to: T.self)
+}
+
+public func dlclose(_ handle: DLHandle) throws {
+    try handle.close()
+}
+
+#if !os(Windows)
+public func dlerror() -> String? {
```

Comment:
On Windows you can implement dlerror using GetLastError + FormatMessage. Example here: https://docs.microsoft.com/en-us/windows/desktop/debug/retrieving-the-last-error-code